validator-info/src/validator_info.rs

```diff
@@ -0,0 +1,506 @@
+use bincode::deserialize;
+use clap::{
+    crate_description, crate_name, crate_version, App, AppSettings, Arg, ArgMatches, SubCommand,
+};
+use serde_derive::{Deserialize, Serialize};
+use serde_json::{Map, Value};
+use solana_client::rpc_client::RpcClient;
+use solana_config_api::{config_instruction, config_instruction::ConfigKeys, ConfigState};
+use solana_sdk::message::Message;
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::signature::{read_keypair, Keypair, KeypairUtil};
+use solana_sdk::transaction::Transaction;
+use std::error;
+use std::process::exit;
+
+pub const MAX_SHORT_FIELD_LENGTH: usize = 70;
+pub const MAX_LONG_FIELD_LENGTH: usize = 300;
+pub const MAX_VALIDATOR_INFO: u64 = 570;
+pub const JSON_RPC_URL: &str = "https://api.testnet.solana.com/";
+
+// Config account key: Va1idator1nfo111111111111111111111111111111
+pub const REGISTER_CONFIG_KEY: [u8; 32] = [
+    7, 81, 151, 1, 116, 72, 242, 172, 93, 194, 60, 158, 188, 122, 199, 140, 10, 39, 37, 122, 198,
+    20, 69, 141, 224, 164, 241, 111, 128, 0, 0, 0,
+];
+
+solana_sdk::solana_name_id!(
+    REGISTER_CONFIG_KEY,
+    "Va1idator1nfo111111111111111111111111111111"
+);
+
+#[derive(Debug, Deserialize, Serialize)]
+struct ValidatorInfo {
+    info: String,
+}
+
+impl ConfigState for ValidatorInfo {
+    fn max_space() -> u64 {
+        MAX_VALIDATOR_INFO
+    }
+}
+
+// Return an error if a pubkey cannot be parsed.
+fn is_pubkey(string: String) -> Result<(), String> {
+    match string.parse::<Pubkey>() {
+        Ok(_) => Ok(()),
+        Err(err) => Err(format!("{:?}", err)),
+    }
+}
+
+// Return an error if a url cannot be parsed.
+fn is_url(string: String) -> Result<(), String> {
+    match url::Url::parse(&string) {
+        Ok(url) => {
+            if url.has_host() {
+                Ok(())
+            } else {
+                Err("no host provided".to_string())
+            }
+        }
+        Err(err) => Err(format!("{:?}", err)),
+    }
+}
+
+// Return an error if url field is too long or cannot be parsed.
+fn check_url(string: String) -> Result<(), String> {
+    is_url(string.clone())?;
+    if string.len() > MAX_SHORT_FIELD_LENGTH {
+        Err(format!(
+            "url longer than {:?}-byte limit",
+            MAX_SHORT_FIELD_LENGTH
+        ))
+    } else {
+        Ok(())
+    }
+}
+
+// Return an error if a validator name is longer than the max length.
+fn check_name_length(string: String) -> Result<(), String> {
+    if string.len() > MAX_SHORT_FIELD_LENGTH {
+        Err(format!(
+            "validator name longer than {:?}-byte limit",
+            MAX_SHORT_FIELD_LENGTH
+        ))
+    } else {
+        Ok(())
+    }
+}
+
+// Return an error if invalid keybase id is submitted.
+fn is_keybase_id(string: String) -> Result<(), String> {
+    let mut invalid = string.len() != 16;
```

Comment:
A keybase user id does not need to be exactly 16 characters, mine is 5 ([mvines](https://keybase.io/mvines)].  If one adds a PGP key to their keybase account, that is 16 characters, but a GPG key is not mandatory for an account and it looks like it's hard to search for a user based on a PGP key with the keybase APIs (whereas searching by user id is trivial).